### ISSUE TITLE
Lots of Numba warnings when measuring correlations

### ISSUE THREAD

**Issue by andreufont** [igmhub/picca#684]
On a fresh new installation of Picca at NERSC, following the instructions here, I get several warnings from numba (see below for full message).

I don't know much (anything) about numba, so it would be great to have some help here.
```
  warnings.warn(errors.NumbaWarning(warn_msg,
/global/homes/f/font/.conda/envs/picca_plots/lib/python3.8/site-packages/numba-0.48.0-py3.8-linux-x86_64.egg/numba/object_mode_passes.py:187: NumbaDeprecationWarning: 
Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

File "../../../../../../homes/f/font/.conda/envs/picca_plots/lib/python3.8/site-packages/picca-4.0-py3.8.egg/picca/xcf.py", line 168:
<source missing, REPL/exec in use?>

  warnings.warn(errors.NumbaDeprecationWarning(msg,
/global/homes/f/font/.conda/envs/picca_plots/lib/python3.8/site-packages/numba-0.48.0-py3.8-linux-x86_64.egg/numba/object_mode_passes.py:177: NumbaWarning: Function "fill_dmat" was compiled in object mode without forceobj=True, but has lifted loops.
```

**Comment by londumas** [comment]
Is it related to issue https://github.com/igmhub/picca/issues/634 ?

**Comment by andreufont** [comment]
It might be. That was an actual error, and it was "fixed" by specifying a particular version of numba, but clearly it didn't solve all the issues.

**Comment by julianbautista** [comment]
These errors are happening because the functions under @jit cannot be compiled in nopython mode. 
See more details here: http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

I tried to modify the code to make it work in nopython mode but I failed. 
So I go with the easy solution for now. 
As recommended I've added:
`@jit(forceobj=True)`
This removes the error messages. 

However, adding forceobj=True is almost identical to not having the `@jit`.  
I did a quick test and the performance is indeed the same. I will keep the jit though for now.

**Comment by iprafols** [comment]
I'm not sure about the performance being the same. The Travis test on the PR #731 failed due to the tests being too long so we need to find a different solution

**Comment by julianbautista** [comment]
Ok.. so it's more complicated than I thought. I'm reverting it back.

**Comment by julianbautista** [comment]
I found a way to code the correlation function with the nopython=True option. 
Not only this removes the warning messages but it also improves the performance of the auto-correlation (cf.py) by a factor of ~**3**!
I created the branch faster_cf. In cf.py, for example, I added `compute_xi_forest_pairs_fast` to be compared with `compute_xi_forest_pairs`.
I will apply similar changes to the other functions that use numba.

**Comment by andreufont** [comment]
Not bad! Have you checked that you reproduce exactly the same results? If so, that's a great improvement!

**Comment by iprafols** [comment]
That's great! If we get exactly the same results I would go ahead and make it the only branch!

**Comment by julianbautista** [comment]
Yes, well, travis is checking that for me !

**Comment by andreufont** [comment]
Would the changes affect the other parts of the code, like the computation of the distortion matrix, metal matrices or wick covariances? Those are the slowest parts of the code, I believe.

It might be worth re-running DR16 before making this the new master, what do you think?

Cheers,
Andreu


________________________________

**Comment by julianbautista** [comment]
That's what I'm working on. There are similar changes that need to be done
to
- distortion matrix
- metal distortion matrix
- all above for the cross correlation

Ok, I could re-run DR16 before merging.

**Comment by iprafols** [comment]
I agree that rerunning DR16 would be nice to make sure the improvement in time escalates properly. We can also use this excuse to finally sort a folder with all the tests together

**Comment by julianbautista** [comment]
I've updated the distortion matrix of the auto. The improvement in time is not as great. 
Here are the times I get running on DR16 Stripe 82 (the tutorial). 

Master
Time spent in picca_cf = 3.3926934679349263 minutes
Time spent in picca_dmat = 1.587285614013672 minutes

Branch faster_cf
Time spent in picca_cf = 0.9995701789855957 minutes
Time spent in picca_dmat = 1.0419551571210226 minutes

The metal matrix are not coded in numba for some reason.

**Comment by julianbautista** [comment]
This is now solved in master with #740
